### DISCUSSION TITLE
Apparent temperature calculation in Open Meteo

### DISCUSSION THREAD

**Discussion by darrenhinderer** [open-meteo/open-meteo#651] [General]
Last weekend I was noticing that the apparent temperature returned from the API seemed especially low compared to what it felt like to me and I started looking into the related open meteo code to understand why.

I found a number of surprises [here](https://github.com/open-meteo/open-meteo/blob/main/Sources/App/Helper/Meteorology.swift#L115), and appreciated that it linked to an apparent temperature calculator which I then used to do some investigation into the related code.

The first thing I want to mention is that the function receives a float argument for humidity which the api does not return, so I can't make my own apparent temp calculation using what is available from the api without losing some small amount of accuracy.

Next I want to point out that the calculator uses a formula from [this paper](http://www.bom.gov.au/jshess/docs/1994/steadman.pdf) and I can't tell if the author of the open meteo code has read that or not. I don't claim to understand the paper but I have tried to read through it.

The first line of the calculation alone is pretty fun!
`let windspeed_2m = windspeed_10m * 0.75`

This isn't necessarily a poor estimate, but it was surprising way to kick things off which could have big implications for the rest of the formula. I also noticed in the Steadman paper he said the effect of wind on people is "barely half" of what is measured at at 10m.

The next line also packs a punch
`let e = relativehumidity_2m/100.0 * 6.105 * exp( 17.27 * temperature_2m / ( 237.7 + temperature_2m ) )`

It took me a while to unpack this but I can now tell you that the paper starting with 6.105 is a formula meant to approximate the vapor pressure of water.  Multiplying the humidity times this formula converts it to the appropriate pressure unit for the final calculation.

But the approximation technique seems to best match up to "Tetens" here which uses slightly different numbers.
https://www.omnicalculator.com/chemistry/vapour-pressure-of-water

I can't tell if the formula having different numbers here would be big enough to notice or if what is used is maybe a better approximation?

And the last piece before the result is calculated is the energy absorbed by the body
` let Q = max(0, 0.1*((shortware_radiation ?? 550) - 550.0))`

Here shortware is a typo, it is the shortwave radiation used to do the calculation. After reading the Steadman paper I'm still befuddled by this code. Why is shortwave multiplied by ".1"? Why is 550 the minimum shortwave value to cause an increase to this beyond zero? 

When I looked at shortwave values where I live in the heat of summer I see them just exceeding 900 and going down to zero. 

The paper says possible values for Q (in Australia) ranged from -40 to 130. It also stated that the energy absorbed can cause an increase in apparent temperature of up to 8 degrees and a decrease -2 degrees when there is heat loss and no energy absorbed.  As the code is written, the possible range of values for Q seems to be more along the lines of 0 to 45.

Finally the last line just puts things together into the Steadman formula, in the only line I didn't find confusing giving the premise
`let AT = temperature_2m + 0.348 * e - 0.70 * windspeed_2m + 0.70 * (Q/(windspeed_2m+10)) - 4.25`

Thanks for reading and I hope this leads to some discussion and possibly improvement to the current apparent temperature calculation.

**Comment by patrick-zippenfenig** [comment]
Hi, Thanks for your in-depth analysis! 

> The first thing I want to mention is that the function receives a float argument for humidity which the api does not return, so I can't make my own apparent temp calculation using what is available from the api without losing some small amount of accuracy.

Do you refer to relative humidity? That would be included in the API output.

> `let windspeed_2m = windspeed_10m * 0.75`

`0.75` is based on [logarithmic wind scale](https://github.com/open-meteo/open-meteo/blob/main/Sources/App/Helper/Meteorology.swift#L156). The FAO uses the same value. Technically, it would be `0.74778485`.

> `let e = relativehumidity_2m/100.0 * 6.105 * exp( 17.27 * temperature_2m / ( 237.7 + temperature_2m ) )`

Yes, this is an approximation for water vapour. There are different approximations available. Some include atmospheric pressure. I do not think that a different approximation makes a noticeable difference.

> ` let Q = max(0, 0.1*((shortware_radiation ?? 550) - 550.0))`

Don't worry about the typo. The Swift compiler would never compile it with an unreferenced variable. IIRC `0.1` is the absorption rate of a human body. Technically you need to get the surface area of a human body and also account for the solar inclination. A tilted human body towards the sun absorbs more energy. 

Arguably the **550 watts maximum** shortwave radiation value is arbitrary. A sunny day in summer may have 1100 watts, but using this value would result in very high apparent temperatures. This might be correct, if you would stand in the sun all day, but otherwise it just forecasts very high values. I added this limit, after users complained about unreasonably high values.

Please feel free to recreate the results, test the formula and make suggestions. A different windspeed reduction from 10 to 2 meters could be considered.

**Reply by patrick-zippenfenig** [reply]
Fixed typo in https://github.com/open-meteo/open-meteo/commit/0c32508c03ff832f03f867b9f16efc2622ca58ee

**Reply by darrenhinderer** [reply]
Thanks Patrick!

What I was referring to about the relative humidity in the API was that it is available only as an integer. Do you think it could be made available with the same decimal precision used here?

Thanks for explaining the rest of the thought process of the algorithm. 

In my case the apparent temperature seeming too cold compared to what it felt like to me was because - the wind value was much higher than what I felt and the sunshine wasn't strong enough to get past the threshold value.

I will continue to test this out and get back to you if I find any solid improvement, but it would help to have the same precision in the API for humidity so I can compare things easier on my end.

**Reply by patrick-zippenfenig** [reply]
Hi, the decimal point precision for relative humidity or other variables should not make much of a difference. If you use the automatically generated Python code, that uses FlatBuffers underneath, you will get all weather data in full precision without any rounding. Rounding is only applied to JSON and CSV outputs as number need to be converted to ASCII strings.

**Reply by darrenhinderer** [reply]
Oh I'm not sure what that all means because I'm just a non-commercial api user right now, building out a weather app for myself.

Maybe I can dive into building the api source later. Thanks again for the help!
